core/src/main/java/org/elasticsearch/cluster/routing/allocation/RebalanceDecision.java

```diff
@@ -77,19 +133,52 @@ public Decision getCanRebalanceDecision() {
         return nodeDecisions;
     }
 
-    private static String produceFinalExplanation(final Type finalDecisionType, final String assignedNodeId, final float threshold) {
-        final String finalExplanation;
-        if (assignedNodeId != null) {
-            if (finalDecisionType == Type.THROTTLE) {
-                finalExplanation = "throttle moving shard to node [" + assignedNodeId + "], as it is " +
-                                       "currently busy with other shard relocations";
+    @Override
+    public String getFinalExplanation() {
+        String explanation;
+        if (fetchPending) {
+            explanation = "cannot rebalance because information about existing shard data is still being retrieved from " +
+                              "some of the nodes, otherwise allocation may prematurely rebalance a shard to a node that will be soon" +
+                              "be imbalanced once shard data retrieval is completed";
+        } else if (canRebalanceDecision.type() != Type.YES) {
+            explanation = "cannot rebalance because rebalancing is not allowed";
+        } else {
+            if (getAssignedNodeId() != null) {
+                if (getFinalDecisionType() == Type.THROTTLE) {
+                    explanation = "throttled on rebalancing shard to another node";
+                } else {
+                    explanation = "can rebalance shard";
+                }
             } else {
-                finalExplanation = "moving shard to node [" + assignedNodeId + "] to form a more balanced cluster";
+                explanation = "cannot rebalance because no other node exists where moving the shard " +
+                                  "to it would form a more balanced cluster";
             }
-        } else {
-            finalExplanation = "cannot rebalance shard, no other node exists that would form a more balanced " +
-                                   "cluster within the defined threshold [" + threshold + "]";
         }
-        return finalExplanation;
+        return explanation;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        super.toXContent(builder, params);
+        builder.field("current_weight", currentWeight);
+        builder.startObject("can_rebalance_decision");
+        {
+            builder.field("final_decision", canRebalanceDecision.type().toString());
+            canRebalanceDecision.toXContent(builder, params);
+        }
+        builder.endObject();
+        if (nodeDecisions != null) {
+            builder.startObject("nodes");
+            {
+                List<String> nodeIds = new ArrayList<>(nodeDecisions.keySet());
+                Collections.sort(nodeIds);
```

Comment:
All the decision objects sort the node decision map by node id as a point of consistency.  We can chose another sort criteria, such as putting all NO decisions up top so they are seen first.  WDYT?